urnai/base/savable.py

```diff
@@ -0,0 +1,196 @@
+        #start_time = time.time()
+        
+        if not self.threaded_saving:
+            self.save_pickle(savepath)
+        else:
+            self.threaded_pickle_save(savepath)
+
+        self.save_extra(savepath)
+        
+        #end_time = time.time()
+        '''
+        rp.report('It took {} seconds to save {} object!'
+		.format(end_time - start_time, self.__class__.__name__),
+        verbosity_lvl=2)
+        '''
+
+    def threaded_pickle_save(self, savepath):
+        """
+        This method saves pickleable
+        stuff in a separate thread
+        """
+        #rp.report('THREADED Saving {} object...'
+		#.format(self.__class__.__name__), verbosity_lvl=1)
+		
+        p = Process(target=self.save_pickle, args=(savepath,))
+        p.start()
+
+    def load(self, loadpath):
+        """
+        This method loads pickle objects
+        and extra stuff needed
+        """
+        self.load_pickle(loadpath)
+        self.load_extra(loadpath)
+
+    def get_pickleable_attributes(self):
+        """
+        This method returns a list of pickeable attributes.
+		If you wish to blocklist one particular pickleable attribute, put it
+        in self.attr_block_list as a string.
+        """
+        if not hasattr(self, 'attr_block_list') or self.attr_block_list is None:
+            self.attr_block_list = []
+
+        full_attr_list = [attr for attr in dir(self) if not attr.startswith('__')
+                          and not callable(getattr(self, attr))
+                          and attr not in self.attr_block_list
+                          and attr != 'attr_block_list']
+        pickleable_list = []
+
+        for key in full_attr_list:
+            try:
+                with tempfile.NamedTemporaryFile() as tmp_file:
+                    pickle.dump(getattr(self, key), tmp_file)
+                    tmp_file.flush()
+                
+                pickleable_list.append(key)
+                
+            except pickle.PicklingError:
+                continue
+            
+            except TypeError as te:
+                if "can't pickle" in str(te):
+                    continue
+                if 'cannot pickle' in str(te):
+                    continue
+                else:
+                    raise
+            
+            except NotImplementedError as nie:
+                if (str(nie) == 
+                'numpy() is only available when eager execution is enabled.'):
+                    continue
+                else:
+                    raise
+            
+            except AttributeError as ae:
+                if ("Can't pickle" in str(ae) or 
+                "object has no attribute '__getstate__'"):
+                    continue
+            
+            except ValueError as ve:
+                if 'ctypes objects' in str(ve):
+                    continue
+                else:
+                    raise
+
+        return pickleable_list
+
+    def get_pickleable_dict(self):
```

Comment:
It should be a generalized method and then a class for the pickle format that would contain the code for that specific format.